### DISCUSSION TITLE
joi .or equivalent?

### DISCUSSION THREAD

**Discussion by pke** [colinhacks/zod#1428] [Q&A] [answered]
```
Joi.object({
  title: Joi.string(),
  text: Joi.string(),
  type: Joi.string().valid("error", "warn", "info", "success").failover("info"),
  duration: Joi.number().positive().default(0),
}).or("text", "title")
```

Here both title and text can be specified but at least on must be specified. How would that be described in zog?

**Comment by scotttrinh** [comment, marked as answer]
You have a couple of options, depending on how you care about the type of the output.

1. This is a union with three members: title; text; title and text;
2. Both are optional and you write a runtime constraint with `refine` that checks that at least one is not `undefined`

Union:
```ts
const typeSchema = z.enum(["error", "warn", "info", "success"]);
const baseSchema = z.object({
  type: typeSchema.optional().default(typeSchema.enum.info),
  duration: z.number().nonnegative().optional().default(0),
});
z.union([
  baseSchema.extend({
    title: z.string(),
  }),
  baseSchema.extend({
    text: z.string(),
  }),
  baseSchema.extend({
    title: z.string(),
    text: z.string(),
  }),
]);
```

refine:
```ts
const typeSchema = z.enum(["error", "warn", "info", "success"]);
z.object({
  title: z.string().optional(),
  text: z.string().optional(),
  type: typeSchema.optional().default(typeSchema.enum.info),
  duration: z.number().nonnegative().optional().default(0),
}).refine(({ title, text }) => title !== undefined || text !== undefined);
```

**Reply by pke** [reply]
Thanks for the quick answer! So everything in zod is requried by default (in joi its optional by default)?
I like the `refine` way better, as its closer to the `joi` way and easier to read imho. I could even be refined (pun) to a `or` function that takes keys of validation and performs the code you showed there:

```ts
z.prototype.or = (keys: string[]) => (
  this.refine(schema => keys.some(key => schema[key]))
)
```
